Adds DLPack support

emcastillo opened the pull request:
Partially Fixes #55090
Depends on #55365

Inspired by https://github.com/dmlc/dlpack/issues/57#issuecomment-774482973

Questions, in PyTorch we can't create streams or easily synchronize them from just an integer. Should we add an [`ExternalStream`](https://docs.cupy.dev/en/stable/reference/generated/cupy.cuda.ExternalStream.html) object like the one we have in CuPy? 

TODO: Add tests

Would like some feedback as this design needs quite a few iterations
@rgommers @leofang

cc @mruberry @rgommers @pmeier @asmeurer @leofang @AnirudhDagar @asi1024 @emcastillo @kmaehashi @heitorschueroff

Commit:
Add support for dlpack to `torch.tensor`

rgommers commented:
> [](https://github.com/dmlc/dlpack/issues/57#issuecomment-774482973)
> 
> Questions, in PyTorch we can't create streams or easily synchronize them from just an integer. Should we add an [`ExternalStream`](https://docs.cupy.dev/en/stable/reference/generated/cupy.cuda.ExternalStream.html) object like the one we have in CuPy?

It shouldn't require a public object I think - I suspect that'd be a bigger discussion (not sure though, @mruberry?). 

Looking at the CuPy version, I think you want a private version (implemented in C++ and with Python bindings so you can use it in `Tensor.__dlpack__`) of:
```python
class ExternalStream(BaseStream):

    """CUDA stream.
    This class allows to use external streams in CuPy by providing the
    stream pointer obtained from the CUDA runtime call.
    The user is in charge of managing the life-cycle of the stream.
    Args:
        ptr (intptr_t): Address of the `cudaStream_t` object.
    Attributes:
        ~Stream.ptr (intptr_t): Raw stream handle.
    """

    def __init__(self, ptr):
        self.ptr = ptr
```
Is that about what you were thinking?

rgommers commented on torch/_tensor.py:
```diff
@@ -1012,6 +1012,34 @@ def __torch_function__(cls, func, types, args=(), kwargs=None):
+
+        Args:
+            stream (integer or None): Object that represents a stream and provides
+                a `synchronize` method. Optional.
```
Oh wait, given the below I think you are extending what's in the array API standard on purpose? What is the use case here? Presumably the caller is coming from another library, would they create/own a PyTorch `Stream` object and work with that instead of with their own stream?

emcastillo replied:
I didn't understand the situation correctly :), I didn't check the actual array API documentation but only the data exchange section which didn't explicitly specify what the stream was, so I went forward with this assumption for a first try.
Will fix it now that I got it! thanks!

rgommers commented on torch/_tensor.py:
```diff
@@ -1012,6 +1012,34 @@ def __torch_function__(cls, func, types, args=(), kwargs=None):
+                a `synchronize` method. Optional.
+        """
+        if isinstance(stream, torch.cuda.Stream) or hasattr(stream, 'synchronize'):
+            stream.synchronize()
```
This shouldn't be necessary, but maybe the right API is missing at the Python level here to do this asynchronously? From https://github.com/dmlc/dlpack/issues/57#issuecomment-753696812:

_"In cases where both sides uses their own stream, async exchange can still be done by stream dependency queing"_:
```c
// event can also be created on the fly, or create a synchronizer object and cache it.
// We could build auxiliary function that can be called from python side if that helps the frameworks
void PushStreamDep(cudaStream_t src, cudaStream dst) {
    cudaEvent_t event;
    cudaEventCreate(&event);
    cudaEventRecord(&event ,src);
    cudaStreamWaitForEvent(dst, event);
   cudaEventDestroy(&event);
}
```

rgommers commented on torch/_tensor.py:
```diff
@@ -1012,6 +1012,34 @@ def __torch_function__(cls, func, types, args=(), kwargs=None):
+        """
+        if isinstance(stream, torch.cuda.Stream) or hasattr(stream, 'synchronize'):
+            stream.synchronize()
+        elif stream is not None and type(stream) is int:
```
It would be good to special-case `-1`, since it's the sentinel value for saying explicitly "do not synchronize".

rgommers commented on torch/utils/dlpack.py:
```diff
@@ -25,3 +13,34 @@
+        # stream
+        if device[0] in (2, 10):
+            stream = torch.cuda.stream.current_stream('cuda:{}'.format(device[1]))
+            # Should we pass an id? or the producer can accept a stream object
```
This should be the stream pointer as an integer.

rgommers commented on torch/utils/dlpack.py:
```diff
@@ -25,3 +13,34 @@
+        if device[0] in (2, 10):
+            stream = torch.cuda.stream.current_stream('cuda:{}'.format(device[1]))
+            # Should we pass an id? or the producer can accept a stream object
+            dlpack = ext_tensor.__dlpack__(stream=stream.cuda_stream)
```
This doesn't handle CPU or other non-CUDA/ROCm devices. I think this should be:

```
        stream = None
        if device[0] in (2, 10):
            stream = torch.cuda.stream.current_stream('cuda:{}'.format(device[1])).cuda_stream
 
        dlpack = ext_tensor.__dlpack__(stream=stream)
```

`Stream.cuda_stream` seems to be the integer stream number and looks like a public attribute, but it's not documented as such. Not sure what's going on there.

rgommers reviewed (commented):
Thanks @emcastillo, looks like a great start.

emcastillo commented:
Regarding the `ExternalStream`, I will send a PR to implement it in PyTorch so we can use it here.
Thanks!

mruberry commented:
> > Questions, in PyTorch we can't create streams or easily synchronize them from just an integer. Should we add an [`ExternalStream`](https://docs.cupy.dev/en/stable/reference/generated/cupy.cuda.ExternalStream.html) object like the one we have in CuPy?
> 
> It shouldn't require a public object I think - I suspect that'd be a bigger discussion (not sure though, @mruberry?).

Keeping things private to start, if possible, is always preferable as it gives us more flexibility in the future. If there's a compelling reason to make it public we can always do so, of course, but you'll have to educate me ;)

Commit:
Use `ExternalStream`

emcastillo commented:
@rgommers I updated the PR after #59527 was merged :).
I tried to address all your concerns, can I get a 2nd review, please?
I also tried to implement the two streams synchronization but did it python side instead of C++. 
I didn't think that increasing the libtorch API with a function that would be used only in this case was worth it if it could be done python side.

Thank you!

leofang commented:
@kmaehashi and I would like to raise this discussion: as demonstrated in @emcastillo's current design, `from_dlpack()` can support both pycapsule objects (for backward compatibility if a library has supported DLPack under this name) and any protocol-complaint object that comes with `__dlpack__` and `__dlpack_device__`. The question is if this is OK to everyone.

rgommers commented:
> @kmaehashi and I would like to raise this discussion: as demonstrated in @emcastillo's current design, `from_dlpack()` can support both pycapsule objects (for backward compatibility if a library has supported DLPack under this name) and any protocol-complaint object that comes with `__dlpack__` and `__dlpack_device__`. The question is if this is OK to everyone.

Thanks for bringing this up @leofang. This seems fine to me, since it's a superset of what's in the array API standard, so there's no conflict. Most libraries will support a superset for other functionality as well, for historical or other reasons.

I would recommend that if this is done, the documentation emphasis that the capsule approach is there only for convenience to support libraries that support the old-style `to_dlpack` and not yet `__dlpack__`. Because the capsule approach is a bit less safe (capsule may not be consumed more than once), takes an extra line of code, and the `stream` parameter cannot be used.

rgommers reviewed (commented):
Thanks @emcastillo. 

> I also tried to implement the two streams synchronization but did it python side instead of C++.
I didn't think that increasing the libtorch API with a function that would be used only in this case was worth it if it could be done python side.

I agree, this makes sense. Having `ExternalStream` available makes this change quite nice and small.

emcastillo commented on torch/_tensor.py:
```diff
@@ -1009,6 +1009,47 @@ def __torch_function__(cls, func, types, args=(), kwargs=None):
+            # from a given pointer
+            raise TypeError('stream must be ``int`` or ``none``')
+        elif stream is not None and stream != -1:
+            if self.device.type in ('cuda', 'rocm'):
```
you are absolutely right, I got confused for a brief moment with torch and dlpack device names 😂,, I know it doesn't make much sense though!

mruberry replied:
You are correct that the device type is 'cuda' even when using a ROCm device.

Check out https://pytorch.org/docs/master/notes/hip.html#checking-for-hip

leofang replied:
I think mocking AMD hardware as an NVIDIA one is an implementation-dependent choice that should not be propagated to `device.type`. IMHO `device.type` should reflect the actual hardware in use.

emcastillo commented on torch/_tensor.py:
```diff
@@ -1009,6 +1009,47 @@ def __torch_function__(cls, func, types, args=(), kwargs=None):
+        # CPU = 1 CPU_PINNED = 3 OPENCL = 4 VULKAN = 7
+        # METAL = 8 VPI = 9
+        dlpack_ids = {'cpu': 1, 'cuda': 2, 'rocm': 10}
+        idx = self.device.index if self.device.index is not None else 0
```
This is a bit out-of-scope but if we were to support these other devices, how would the stream support work?
Should it be ignored in environments where a stream does not make any sense?

leofang replied:
I think what @rgommers meant is to change the return type of this function:
```python
def __dlpack_device__(self) -> Tuple[enum.IntEnum, int]:
```
> This is a bit out-of-scope but if we were to support these other devices, how would the stream support work?
> Should it be ignored in environments where a stream does not make any sense?

For `__dlpack_device__` whether a device has the concept of stream/queue doesn't matter. For `__dlpack__` stream can be `Any`: 
https://data-apis.org/array-api/latest/API_specification/array_object.html#dlpack-self-stream-none

emcastillo commented on torch/utils/dlpack.py:
```diff
@@ -25,3 +13,37 @@
 """)
+
+
+def from_dlpack(ext_tensor) -> torch.Tensor:
```
Thanks!

Commit:
Review changes and test

emcastillo commented:
I think I addressed all the review concerns, also added some small tests to verify the behavior.
Can I take another look?
This should be close to landing, but maybe some small fixes or tweaks might be required.

Commit:
More stream checks

rgommers commented:
There's a bunch of test failures. Not sure about the `xla` ones.

For the `__torch_function__` ones, the dunder methods need a snippet like this at the top of the file:
```python
        if has_torch_function_unary(self):
            return handle_torch_function(Tensor.__dlpack__, (self,), self, stream)
```

One other thing to discuss: `from_dlpack` is left in the `torch.utils.dlpack` namespace here, but according to the array API standard we'd want it in the main namespace. I think it fits there too, next to `from_numpy` and (soon) `frombuffer`.

Commit:
Review fixes

emcastillo commented:
I think the failures are unrelated?

rgommers commented:
> I think the failures are unrelated?

They all are related (at the first 6 CI jobs I checked):

<details>

```
======================================================================
ERROR [0.001s]: test_dlpack_protocol_conversion_cpu_uint8 (__main__.TestTorchDeviceTypeCPU)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_device_type.py", line 397, in instantiated_test
    result = test_fn(self, *args)
  File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_device_type.py", line 764, in dep_fn
    return fn(slf, device, *args, **kwargs)
  File "test_torch.py", line 6960, in test_dlpack_protocol_conversion
    z = from_dlpack(x)
  File "/opt/conda/lib/python3.6/site-packages/torch/utils/dlpack.py", line 39, in from_dlpack
    device = ext_tensor.__dlpack_device__()
  File "/opt/conda/lib/python3.6/site-packages/torch/_tensor.py", line 1067, in __dlpack_device__
    raise ValueError('Unknown device type {} for Dlpack'.format(device_type))
ValueError: Unknown device type cpu for Dlpack


======================================================================
ERROR [0.000s]: test_dlpack_protocol_conversion_cpu_uint8 (__main__.TestTorchDeviceTypeCPU)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "C:\actions-runner\_work\pytorch\pytorch\build\win_tmp\build\torch\testing\_internal\common_device_type.py", line 397, in instantiated_test
    result = test_fn(self, *args)
  File "C:\actions-runner\_work\pytorch\pytorch\build\win_tmp\build\torch\testing\_internal\common_device_type.py", line 764, in dep_fn
    return fn(slf, device, *args, **kwargs)
  File "test_torch.py", line 6960, in test_dlpack_protocol_conversion
    z = from_dlpack(x)
  File "C:\actions-runner\_work\pytorch\pytorch\build\win_tmp\build\torch\utils\dlpack.py", line 39, in from_dlpack
    device = ext_tensor.__dlpack_device__()
  File "C:\actions-runner\_work\pytorch\pytorch\build\win_tmp\build\torch\_tensor.py", line 1066, in __dlpack_device__
    if device_type not in DLPackIds:
  File "c:\jenkins\miniconda3\lib\enum.py", line 352, in __contains__
    raise TypeError(
TypeError: unsupported operand type(s) for 'in': 'str' and 'EnumMeta'


======================================================================
ERROR [0.001s]: test_Tensor___dlpack__ (__main__.TestTorchFunctionOverride)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "test_overrides.py", line 630, in test
    ret = func(*func_args)
  File "/opt/conda/lib/python3.6/site-packages/torch/_tensor.py", line 1034, in __dlpack__
    return handle_torch_function(Tensor.__dlpack__, (self,), self, stream)
  File "/opt/conda/lib/python3.6/site-packages/torch/overrides.py", line 1308, in handle_torch_function
    result = overloaded_arg.__torch_function__(public_api, types, args, kwargs)
  File "test_overrides.py", line 212, in __torch_function__
    return HANDLED_FUNCTIONS_SUB[func](*args, **kwargs)
  File "test_overrides.py", line 301, in wrapped
    return f(*args, **kwargs)
TypeError: <lambda>() takes 1 positional argument but 2 were given

```

</details>

Commit:
fix tests

emcastillo commented:
> They all are related (at the first 6 CI jobs I checked):

This is embarrassing 😅, I was looking at the summaries in the posted comment, I couldn't access the actual ci pages, so I spoke too fast. Sorry and thank you for double-checking it!

Commit:
fix tests

Commit:
filter xla

rgommers commented:
> I was looking at the summaries in the posted comment, I couldn't access the actual ci pages, so I spoke too fast. Sorry and thank you for double-checking it!

Can you access them now? If not, we should ask for more permissions for you.

rgommers commented on test/test_torch.py:
```diff
@@ -6947,15 +6947,89 @@ def compare_strides(s1, s2, div):
+                assert stream_a.query()
+                assert stream_b.query()
+                z = _from_dlpack(x.__dlpack__(stream_b.cuda_stream))
+                # sync in stream a forces the stream b work to be completed
```
minor: it's not a sync, it's a "wait for event" right? maybe worth making that clear in the comment.

rgommers reviewed (commented):
Looks pretty good to me now. All CI failures seem unrelated (only ROCm is still running at the moment).

The stream tests I'm not 100% sure about, but then again I'm not much of a CUDA programmer - would be nice if someone else could take a good look at those.

leofang commented on test/test_torch.py:
```diff
@@ -6947,15 +6947,89 @@ def compare_strides(s1, s2, div):
+            with torch.cuda.stream(stream):
+                assert stream.query()
+                z = from_dlpack(x)
+                assert not stream.query()
```
Is this line deterministic?

emcastillo replied:
I thought that the query retturns False until synchronization actually happens, but I am not sure now.

leofang commented on torch/_tensor.py:
```diff
@@ -1012,6 +1013,61 @@ def __torch_function__(cls, func, types, args=(), kwargs=None):
+        of the consumer library.
+
+        Args:
+            stream (integer or None): A Python integer representing a pointer
+            to a stream (CUDA or ROCm). `stream` is provided by the consumer
```
Is there a place we can check the rendered doc? I thought from the 2nd line onward we need an indentation?

rgommers replied:
There normally is, in the CI list there's a doc artifact link. However, on this run the doc build failed with an unrelated issue:
```
/var/lib/jenkins/workspace/docs/source/notes/cuda.rst:270: WARNING: Definition list ends without a blank line; unexpected unindent.
```
so no artifact was generated.

mruberry replied:
Rendered docs are now available as a link on the Dr. CI post:

https://pytorch.org/docs/master/dlpack.html?highlight=dlpack#torch.utils.dlpack.to_dlpack

leofang commented on torch/_tensor.py:
```diff
@@ -1012,6 +1013,61 @@ def __torch_function__(cls, func, types, args=(), kwargs=None):
+                # Only synchronize on different streams
+                if stream != torch.cuda.current_stream:
+                    event = torch.cuda.Event()
+                    event.record(stream)
+                    torch.cuda.current_stream().wait_event(event)
```
Should be the other way around: The consumer's stream waits for the producer (PyTorch)'s stream:
```suggestion
                    event.record(torch.cuda.current_stream())
                    stream.wait_event(event)
```

emcastillo replied:
thanks!

leofang commented on torch/_tensor.py:
```diff
@@ -1012,6 +1013,61 @@ def __torch_function__(cls, func, types, args=(), kwargs=None):
+            return handle_torch_function(Tensor.__dlpack_device__, (self,), self)
+        idx = self.device.index if self.device.index is not None else 0
+        device_type = self.device.type
+        if device_type == 'cuda' and torch.version.hip is not None:
+            device_type = 'rocm'
+        elif device_type == 'cpu' and self.is_pinned():
+            device_type = 'cpu_pinned'
```
Question: `device_type` will not be `'cpu'` in PyTorch?

emcastillo replied:
sorry I dont' follow you

leofang replied:
My bad, sorry Emilio. I was being silly early on...

leofang commented on test/test_torch.py:
```diff
@@ -6947,15 +6947,89 @@ def compare_strides(s1, s2, div):
+                z = from_dlpack(x)
+                assert not stream.query()
+            assert stream.query()
+            self.assertEqual(z, x)
```
Question: is `z` guaranteed to be ready when we go out of the scope of a PyTorch stream? If not, then this assert should be moved to within the scope.

emcastillo replied:
when going out of the scope of a stream, I think that all the work queued on it should be done, but I am not sure now.

leofang commented on test/test_torch.py:
```diff
@@ -6947,15 +6947,89 @@ def compare_strides(s1, s2, div):
+            return
+        if device == 'cuda':
+            from torch._C import _from_dlpack
+            x = make_tensor((5,), device, dtype, low=-9, high=9)
+            stream_a = torch.cuda.Stream()
+            stream_b = torch.cuda.Stream()
+            with torch.cuda.stream(stream_a):
+                assert stream_a.query()
+                assert stream_b.query()
+                z = _from_dlpack(x.__dlpack__(stream_b.cuda_stream))
```
Same question here: is `x` guaranteed to be ready when we enter the scope of `stream_a`? I would have expected this snippet instead:
```python
with torch.cuda.stream(stream_a):
    x = make_tensor((5,), device, dtype, low=-9, high=9)
with torch.cuda.stream(stream_b):
    z = from_dlpack(x)
```
But maybe there's some assumptions in PyTorch's stream context that I missed?

emcastillo commented on test/test_torch.py:
```diff
@@ -6947,15 +6947,89 @@ def compare_strides(s1, s2, div):
+    @dtypes(*torch.testing.get_all_dtypes())
+    def test_dlpack_protocol_conversion(self, device, dtype):
+        # DLpack does not explicitly support bool
+        # It does it through uint8 type
```
much better! thanbks :)

Commit:
fixes

emcastillo commented on torch/_tensor.py:
```diff
@@ -1026,6 +1027,61 @@ def __torch_function__(cls, func, types, args=(), kwargs=None):
+                if stream != torch.cuda.current_stream:
+                    event = torch.cuda.Event()
+                    event.record(torch.cuda.current_stream())
+                    torch.cuda.current_stream().wait_event(event)
```
thks!

mruberry commented on test/test_torch.py:
```diff
@@ -6977,15 +6977,64 @@ def compare_strides(s1, s2, div):
+        # DLpack does not explicitly support bool (xref dmlc/dlpack#75)
         # It does it through uint8 type
-        if dtype is torch.bool:
+        if dtype is torch.bool or 'xla' in device:
```
Don't skip the test like this. Use the @onlyOnCPUAndCUDA decorator to run only on CPU and CUDA devices. Remove torch.bool from the set of dtypes passed to the `@dtypes` decorator if the test doesn't want to test bool.

However, what about asserting an error is thrown for bool tensors?

emcastillo replied:
Done, I added a test just for bool tensors :)

mruberry commented on test/test_torch.py:
```diff
@@ -6977,15 +6977,64 @@ def compare_strides(s1, s2, div):
+        if dtype is torch.bool or 'xla' in device:
+            return
+        # Create a stream where the tensor will reside
+        if device == 'cuda':
```
The device string may not be 'cuda'. To check if the device is a cuda device use `torch.device(device).type == 'cuda'`.

mruberry replied:
If this test should ONLY run on CUDA devices then use the `@onlyOnCUDA` decorator

emcastillo replied:
much better, thanks!

mruberry commented on test/test_torch.py:
```diff
@@ -6977,15 +6977,64 @@ def compare_strides(s1, s2, div):
+        # Create a stream where the tensor will reside
+        if device == 'cuda':
+            x = make_tensor((5,), device, dtype, low=-9, high=9)
+            stream = torch.cuda.Stream()
```
Do you need a synchronize call here to ensure `x` has been populated on a CUDA device before its accessed on a different stream?

emcastillo replied:
Here we want to test how the dlpack interface does this synchronization, so we avoid doing it here.

emcastillo replied:
I added a comment to clarify this

mruberry commented on test/test_torch.py:
```diff
@@ -6977,15 +6977,64 @@ def compare_strides(s1, s2, div):
+        if device == 'cuda':
+            from torch._C import _from_dlpack
+            x = make_tensor((5,), device, dtype, low=-9, high=9)
+            stream_a = torch.cuda.Stream()
```
this needs to make sure that x is available before its accessed on a stream

emcastillo replied:
same as above, added a comment, `tensor.__dlpack__` deals with this interstream synchronization as @leofang pointed out in https://github.com/pytorch/pytorch/pull/57110#issuecomment-875270958

mruberry commented on test/test_torch.py:
```diff
@@ -6977,15 +6977,64 @@ def compare_strides(s1, s2, div):
+            stream_a = torch.cuda.Stream()
+            stream_b = torch.cuda.Stream()
+            with torch.cuda.stream(stream_a):
+                z = _from_dlpack(x.__dlpack__(stream_b.cuda_stream))
```
What are the semantics here? x is converted to a dlpack tensor on stream b, and then z is constructed from the dlpack tensor on stream a? In that case, stream a needs to wait for stream b to complete is work

emcastillo replied:
here, `__dlpack__` will add an event to `stream_b` to enforce ordering with `stream_a`

https://github.com/pytorch/pytorch/blob/4e90170386a80278a25e7027a742db66ebf7f035/torch/_tensor.py#L1055-L1060

mruberry commented on torch/_tensor.py:
```diff
@@ -1026,6 +1027,61 @@ def __torch_function__(cls, func, types, args=(), kwargs=None):
+        Args:
+            stream (integer or None): A Python integer representing a pointer
+            to a stream (CUDA or ROCm). `stream` is provided by the consumer
+            to the producer to instruct the producer to ensure that operations
```
How can the "consumer" provide the stream? Doesn't the user provide the stream?

emcastillo replied:
`__dlpack__` is called from `from_dlpack` method and it will pass the current stream as an argument as part of the dlpack specification. Will update the docstring to reflect this!

mruberry commented on torch/_tensor.py:
```diff
@@ -1026,6 +1027,61 @@ def __torch_function__(cls, func, types, args=(), kwargs=None):
 
+    def __dlpack__(self, stream=None):
+        """
+        Creates a dlpack capsule of the current tensor to
```
"Creates a DLPack capsule from the current tensor."

The phrase "DLPack capsule" should be a link to the DLPack documentation.

mruberry commented on torch/_tensor.py:
```diff
@@ -1026,6 +1027,61 @@ def __torch_function__(cls, func, types, args=(), kwargs=None):
+        Creates a dlpack capsule of the current tensor to
+        be exported to other libraries.
+
+        This function will be called from the `from_dlpack` method
```
I think we can get rid of this sentence?

emcastillo replied:
I think is better to leave it as this method is not intended to be called by users, I extended it to note that `from_dlpack` passes the current stream

mruberry replied:
OK

mruberry commented on torch/_tensor.py:
```diff
@@ -1026,6 +1027,61 @@ def __torch_function__(cls, func, types, args=(), kwargs=None):
+            to the producer to instruct the producer to ensure that operations
+            can safely be performed on the array.
+            The pointer must be a positive integer or
+            -1 . If stream is -1 , the value may be used by the consumer to
```
I'm not really sure what -1 means from this description -- is there a documentation pointer?

emcastillo replied:
yes, the dlpack documentation covers this https://github.com/rgommers/array-api/blob/main/spec/API_specification/array_object.md#__dlpack__--streamnone

mruberry commented on torch/_tensor.py:
```diff
@@ -1026,6 +1027,61 @@ def __torch_function__(cls, func, types, args=(), kwargs=None):
+        try:
+            return (DLPackIds[device_type], idx)
+        except KeyError:
+            raise ValueError('Unknown device type {} for Dlpack'.format(device_type))
```
"DLPack"

mruberry commented on torch/_tensor.py:
```diff
@@ -1026,6 +1027,61 @@ def __torch_function__(cls, func, types, args=(), kwargs=None):
+            # Stream pointers in CUDA/ROCm are uniquely numbered and can
+            # be retrieved from their integer value.
+            raise TypeError('stream must be ``int`` or ``none``')
+        elif stream is not None and stream != -1:
```
What's the idea of the stream argument here and why would we want this to automatically synchronize streams for the user?

emcastillo replied:
This is part of the dlpack specification, @leofang and @rgommers can help here :)

mruberry commented on torch/_tensor.py:
```diff
@@ -1026,6 +1027,61 @@ def __torch_function__(cls, func, types, args=(), kwargs=None):
+                    event = torch.cuda.Event()
+                    event.record(torch.cuda.current_stream())
+                    stream.wait_event(event)
+        return torch.to_dlpack(self)
```
So if an integer is provided this uses that integer to acquire an ExternalStream, then syncs that external stream with the current stream, but then the `to_dlpack` call discards the stream operation?

What is this intended to accomplish?

emcastillo replied:
This ensures that the data will be copied at the right moment by forcing an order between the external library stream and pytorch stream.

https://github.com/dmlc/dlpack/issues/57#issuecomment-753696812

mruberry replied:
OK I think I understand this now because the `to_dlpack` operation is not actually a CUDA operation since the memory is shared.

mruberry commented on torch/utils/dlpack.py:
```diff
@@ -25,3 +15,38 @@
+def from_dlpack(ext_tensor: Any) -> torch.Tensor:
+    """from_dlpack(ext_tensor) -> Tensor
+
+    Convers a tensor from a external library into a ``torch.Tensor``
```
"Converts a DLPack capsule into a tensor."

The phrase "DLPack capsule" should be a link to the DLPack documentation

mruberry replied:
This comment still needs to be resolved

rgommers replied:
> The phrase "DLPack capsule" should be a link to the DLPack documentation

There's no such thing yet. DLPack itself only has comments in [dlpack.h](https://github.com/dmlc/dlpack/blob/main/include/dlpack/dlpack.h). `__dlpack__` is part of the array API standard, see [here](https://data-apis.org/array-api/latest/API_specification/array_object.html#dlpack-self-stream-none).

emcastillo replied:
I am fine with adding the link to the API standard, would it be possible to do it in another PR?
Thanks!

mruberry commented on torch/utils/dlpack.py:
```diff
@@ -25,3 +15,38 @@
+    Convers a tensor from a external library into a ``torch.Tensor``
+    by means of the ``__dlpack__`` protocol.
+
+    The tensor will share the memory with the object represented
```
"The tensor and the DLPack capsule share their memory."

Is this true when converting a tensor to a DLPack capsule, too? How does that work if a tensor requires autograd?

mruberry replied:
If converting to a DLPack capsule from a torch tensor is also memory sharing, then I think we should model the operation after `.numpy()`, which does the following when a tensor requires grad:

```
t = torch.tensor((1., 2., 3.), requires_grad=True)
t.numpy()
: RuntimeError: Can't call numpy() on Tensor that requires grad. Use tensor.detach().numpy() instead.
```

See https://github.com/pytorch/pytorch/pull/59790/files for the latest on `.numpy()` semantics. We also don't support tensors with the conjugate bit set, for example.

emcastillo replied:
this is a good point

mruberry commented on torch/utils/dlpack.py:
```diff
@@ -25,3 +15,38 @@
+    The tensor will share the memory with the object represented
+    in the dlpack.
+
+    In order to keep backward compatibility, this function also admits
```
Backward compatibility with what?

emcastillo replied:
old's pytorch `from_dlpack` implementation, previously it took a dlpack capsule object, now we pass a, object with `__dlpack__` protocol, so we accept both objects in order to not break existing code.

This can be changed.

mruberry replied:
Accepting both seems great. I just don't think most people will understand the motivation being driven by BC, so we probably shouldn't mention it.

mruberry commented on torch/utils/dlpack.py:
```diff
@@ -25,3 +15,38 @@
+
+    In order to keep backward compatibility, this function also admits
+    to pass a dlpack capsule object.
+    Note that each dlpack capsule can only be consumed once.
```
What happens if `from_dlpack()` is called on the same dlpack capsule more than once?

emcastillo replied:
something like a double free errror is likely to happen. I think the C layer checks this.

mruberry commented on torch/utils/dlpack.py:
```diff
@@ -25,3 +15,38 @@
+
+    Args:
+        ext_tensor (object with __dlpack__ attribute or dlpack capsule):
+            The tensor from an external library that will be converted
```
"The tensor or DLPack capsule to convert."

mruberry commented on torch/utils/dlpack.py:
```diff
@@ -25,3 +15,38 @@
+        device = ext_tensor.__dlpack_device__()
+        # device is either CUDA or ROCm, we need to pass the current
+        # stream
+        if device[0] in (2, 10):
```
What is the test for the device being 2 or 10? Are those just magic numbers defined by the DLPack specification?

emcastillo replied:
yes, I should change this to actual nemonics, sorry

mruberry commented on test/test_torch.py:
```diff
@@ -6977,15 +6977,64 @@ def compare_strides(s1, s2, div):
+        if dtype is torch.bool or 'xla' in device:
             return
         x = make_tensor((5,), device, dtype, low=-9, high=9)
         z = from_dlpack(to_dlpack(x))
```
It'll be nice if/when NumPy implements the DLPack protocol because we can actually validate that we can import NumPy tensors using the protocol

emcastillo replied:
Left a TODO

mruberry commented on test/test_torch.py:
```diff
@@ -6977,15 +6977,64 @@ def compare_strides(s1, s2, div):
 
+    @skipMeta
+    @dtypes(*torch.testing.get_all_dtypes())
+    def test_dlpack_protocol_conversion(self, device, dtype):
```
We should add a test that compares the DLPack semantics with our .numpy() and from_numpy() semantics. If/when NumPy implements the protocol we could really validate that the behavior is the same

emcastillo replied:
Sorry I don't understand what should be compared here.

emcastillo replied:
Now, I think I understood it,
A test that checks that tensors from dlpack can't be resized, or tensors with gradients can't be exported in the same sense that numpy does, is this correct?

mruberry replied:
Yes: tensors with gradients or the conjugate bit set, I think, can't be exported.

For importing we should verify that the underlying memory is shared by writing to it on both CPU and CUDA.

NumPy arrays can also be non-writable, which we check for on import. I'm not sure what (if any) special properties DLPack capsules have that PyTorch can't emulate.

mruberry reviewed (commented):
Hey @emcastillo! I was just checking in and I left some inline comments on the PR, mostly for clarification about stream behavior and some doc suggestions. I also think we should align the DLPack to/from semantics with our NumPy to/from semantics, too. Looking forward to hearing your thoughts!

leofang commented:
> mostly for clarification about stream behavior and some doc suggestions. 

Hi @mruberry I think many of your questions are answered (and dictated) by the DLPack exchange protocol, so I would suggest to review the spec first:
- main doc is https://data-apis.org/array-api/latest/design_topics/data_interchange.html
- `__dlpack__`: https://data-apis.org/array-api/latest/API_specification/array_object.html#dlpack-self-stream-none
- `__dlpack_device__`: https://data-apis.org/array-api/latest/API_specification/array_object.html#dlpack-device-self
- `from_dlpack`: https://data-apis.org/array-api/latest/API_specification/creation_functions.html#function-from-dlpack

In a nutshell, the DLPack protocol helps establish a correct stream order (hence data dependency) at the exchange boundary.  

> I also think we should align the DLPack to/from semantics with our NumPy to/from semantics, too. 

I am not 100% sure what this means, maybe @emcastillo knows? 🙂 If it's about the behavior of `from_dlpack()`, it is fully defined for CPU/CUDA/ROCm by the Array API standard.

emcastillo commented:
@mruberry Thanks a lot for all the comments!
they are super helpful, I will try to improve the documentation with links to the actual dlpack docs so there is little room for confusion. I agree that a person who is not familiar with dlpack will have a difficult time understanding this implementation right now.
Thanks again!

Commit:
Enhancements

emcastillo commented:
@mruberry @rgommers hey!
I tried to address most of your concerns here :)
Can you please take another look? probably this needs one or two iterations more 😄

Thanks!!

Commit:
lint

Commit:
fix-mypy

Commit:
flake8 fix

mruberry commented:
> @mruberry @rgommers hey!
> I tried to address most of your concerns here :)
> Can you please take another look? probably this needs one or two iterations more 😄
> 
> Thanks!!

Sorry for the long delay, @emcastillo. Looks like this notification was lost. If I don't get back to you in two business days please ping me again! Looking now.

mruberry commented on test/test_torch.py:
```diff
@@ -7042,16 +7042,93 @@ def compare_strides(s1, s2, div):
+        # explicitly wait until x is populated
+        stream = torch.cuda.Stream()
+        with torch.cuda.stream(stream):
+            z = from_dlpack(x)
```
In this case won't the storage be shared so this operation doesn't actually perform any work on the stream?

emcastillo replied:
I changed it to generate an array in another stream, but the tests with streams doesn't make much sense when doing roundtrips to the same library 😣, we might need a mocked `__dlpack__` protocol for testing.

mruberry commented on test/test_torch.py:
```diff
@@ -7042,16 +7042,93 @@ def compare_strides(s1, s2, div):
+    @dtypes(*torch.testing.get_all_dtypes(include_bool=False))
+    def test_dlpack_conversion_with_diff_streams(self, device, dtype):
+        from torch._C import _from_dlpack
+        x = make_tensor((5,), device, dtype, low=-9, high=9)
```
Nit: make_tensor() doesn't require low/high be specified and the particular values don't seem important here

mruberry commented on test/test_torch.py:
```diff
@@ -7042,16 +7042,93 @@ def compare_strides(s1, s2, div):
+    # TODO: increase tests once NumPy supports the `__dlpack__` protocol
+
+    @skipMeta
+    def test_dlpack_export_requires_grad(self):
```
Nice requires grad and conj bit tests

mruberry replied:
Let's add a test for layouts other than strided, too, like csr

emcastillo replied:
added! thanks :)

mruberry commented on torch/_tensor.py:
```diff
@@ -1033,6 +1034,66 @@ def __torch_function__(cls, func, types, args=(), kwargs=None):
+            can safely be performed on the array.
+            The pointer must be a positive integer or
+            -1 . If stream is -1 , the value may be used by the consumer to
+            signal "producer must not perform any synchronization. Optional.
```
Unclosed quote here

mruberry commented on torch/_tensor.py:
```diff
@@ -1033,6 +1034,66 @@ def __torch_function__(cls, func, types, args=(), kwargs=None):
+        stream to this method as part of the specification.
+
+        Args:
+            stream (integer or None): A Python integer representing a pointer
```
OK I think I understand this. Let me try a suggestion:

"An optional Python integer representing a pointer to a CUDA stream. The current stream is synchronized with this stream after the capsule is created, making the capsule safe to access from this stream. If None or -1 is passed then no synchronization is performed." 

Is that accurate?

emcastillo replied:
much better, thanks!

mruberry commented on torch/_tensor.py:
```diff
@@ -1033,6 +1034,66 @@ def __torch_function__(cls, func, types, args=(), kwargs=None):
+        if has_torch_function_unary(self):
+            return handle_torch_function(Tensor.__dlpack__, (self,), self, stream)
+
+        # Some semantics that can prevent tensors from being exported are
```
More precisely, DLPack capsules can't capture all of PyTorch's semantics, so we prohibit exporting tensors that would lose their properties like requires_grad and having the conjugate bit set.

mruberry commented on torch/utils/dlpack.py:
```diff
@@ -25,3 +30,37 @@
+    The tensor will share the memory with the object represented
+    in the dlpack.
+
+    Note that each dlpack capsule can only be consumed once. Otherwise
```
Here and elsewhere we should be careful to use the proper name "DLPack" consistently

mruberry commented on torch/utils/dlpack.py:
```diff
@@ -25,3 +30,37 @@
+    in the dlpack.
+
+    Note that each dlpack capsule can only be consumed once. Otherwise
+    memory errors could happen.
```
I suggest making this a warning:

```
.. warning::
  Only call from_dlpack once per capsule. Its behavior when used on the same capsule multiple times is undefined.
```

mruberry reviewed (commented):
Hey @emcastillo!

This looks pretty good. I made a few inline suggestions but I think it's very close to be doing done. Looking forward to seeing an update and again, sorry about the delay. Ping me when this is ready for another review and that should happen quickly!

emcastillo commented:
Thanks Mike!!!

Will give it another pass during the week and get back to you shortly :)

Commit:
Review changes

emcastillo commented:
@mruberry , sorry for the delay!!
I think the last commit should try to address your comments :)
Can you check it again please? thanks!

mruberry reviewed (approved):
LGTM! Thanks @emcastillo. Just see one unresolved comment suggesting a doc tweak to the "from_dlpack" function.

@rgommers Would you like to take a final look?
@anjali411 are there any updates to exporting tensors with the conj bit we should be aware of?

rgommers reviewed (approved):
This looks great, thanks @emcastillo!

emcastillo commented:
Thank you all for your advice and for taking the time to thoroughly review my horrible initial implementation 😅

mruberry commented:
Doh! Sorry, @emcastillo, looks like this pick up a merge conflict. Would just rebase it and ping me so I can merge it?

emcastillo commented:
@mruberry rebased!

rgommers commented on torch/__init__.py:
```diff
@@ -759,6 +759,8 @@ def compiled_with_cxx11_abi():
 quantized_lstm = torch.ops.aten.quantized_lstm
 quantized_gru = torch.ops.aten.quantized_gru
 
+from torch.utils.dlpack import from_dlpack, to_dlpack
```
Adding some docs for DLPack in gh-70437. I'd forgotten about this. Looked back at my review comment, and it mentioned just `from_dlpack`. I think adding `to_dlpack` to the main namespace was probably unnecessary, given that it shouldn't be used when all libraries support `__dlpack__`. Anyway, no action needed - mostly a comment to self.
